### PR TITLE
Lowering transpose operator

### PR DESCRIPTION
**Pull request by seibert**
This implements the lowering of the graphblas.transpose operator (I know we want to rename this, but will do that in future PR).  This also creates some utility functions to help with fetching / defining external helper methods.

### CODE REVIEW

File: mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#ifndef GRAPHBLAS_GRAPHBLASUTILS_H
+#define GRAPHBLAS_GRAPHBLASUTILS_H
+
+#include <string>
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+
+
```
Comment (paul-tqh-nguyen) [inline mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h:9]: Would it be easier if we used `using namespace mlir` here?

That might let us get rid of the "mlir::" prefixes everywhere.

File: mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/SparseTensor/IR/SparseTensor.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+
+using namespace std;
+using namespace mlir;
+using namespace mlir::sparse_tensor;
+
+// make CSR tensor type
+mlir::RankedTensorType getCSRTensorType(MLIRContext *context, Type valueType)
```
Comment (paul-tqh-nguyen) [inline mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp:13]: I think we can replace `mlir::RankedTensorType` -> `RankedTensorType` because of `using namespace mlir;` above.

File: mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        {-1, -1}, /* 2D, unknown size */
+        valueType,
+        sparse_tensor::SparseTensorEncodingAttr::get(context, dlt, map, ptr, ind));
+
+    return csrTensor;
+}
+
+/// Returns function reference (first hit also inserts into module).
+// from: llvm/llvm-project/mlir/lib/Dialect/SparseTensor/Transforms/SparseTensorConversion.cpp
+static FlatSymbolRefAttr getFunc(mlir::ModuleOp &mod, mlir::Location &loc, StringRef name, TypeRange result,
```
Comment (paul-tqh-nguyen) [inline mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp:32]: I think we can remove the `mlir::` prefixes because of `using namespace mlir;` above. It would help the code look more consistent since we in other places use `MLIRContext` instead of `mlir::MLIRContext`.

I think we can do this throughout this file.

File: mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        moduleBuilder
+            .create<FuncOp>(loc, name,
+                            FunctionType::get(context, operands, result))
+            .setPrivate();
+    }
+    return SymbolRefAttr::get(context, name);
+}
+
+mlir::CallOp callEmptyLike(mlir::OpBuilder &builder, mlir::ModuleOp &mod, mlir::Location loc, mlir::Value tensor) {
+    auto tensorType = tensor.getType();
```
Comment (paul-tqh-nguyen) [inline mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp:49]: While it's nice to use `auto`, I like making the types explicit since it'll be easier to look up in the MLIR doxygen docs what methods are available for a certain value. 

For example, if I knew that `tensorType` was a `mlir::Type`, I could find at https://mlir.llvm.org/doxygen/classmlir_1_1Type.html what methods were available for it. 

There's also some implicit knowledge here in labelling it as `mlir::Type` (instead of `mlir::TensorType`, which is what it would actually always be in this function) telling us that the `getType` method returns a general `mlir::Type` and that we'd have to dynamically cast it to a `mlir::TensorType` if we wanted to use methods specific to `mlir::TensorType`. 
- In short, it's nice to know that the compiler sees this as a `mlir::Type` and not a `mlir::TensorType` so that we'd know we have to dynamically cast it if we wanted to use the methods of `mlir::TensorType`.

A silly hack that I use to determine the type of a variable is to print it (which'll cause a compile-time type error), compile the code, and read the error message of "Expected X, but got Y" to know that the variable's type should be Y. This won't work if it is printable, but that's rarely the case.

File: mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir [hunk 5]
```diff
@@ -0,0 +1,87 @@
+// RUN: graphblas-opt %s | graphblas-opt --graphblas-lower | FileCheck %s
+
+#CSR64 = #sparse_tensor.encoding<{
+  dimLevelType = [ "dense", "compressed" ],
+  dimOrdering = affine_map<(i,j) -> (i,j)>,
+  pointerBitWidth = 64,
+  indexBitWidth = 64
+}>
+ 
+
```
Comment (paul-tqh-nguyen) [inline mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir:10]: Was this whole test manually written? 

If so, it seems like this sort of test would be hard to maintain. Maybe we should make it not necessarily check every single line and have a separate test verify that we get the same behavior once compiled to make sure all the lowering does what we want.

File: mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+// CHECK:             memref.store %[[VAL_44]], %[[VAL_14]]{{\[}}%[[VAL_9]]] : memref<?xi64>
+// CHECK:           }
+// CHECK:           memref.store %[[VAL_42]], %[[VAL_14]]{{\[}}%[[VAL_9]]] : memref<?xi64>
+// CHECK:           return %[[VAL_13]] : tensor<?x?xf64, #sparse_tensor.encoding<{ dimLevelType = [ "dense", "compressed" ], dimOrdering = affine_map<(d0, d1) -> (d0, d1)>, pointerBitWidth = 64, indexBitWidth = 64 }>>
+// CHECK:         }
+
+func @transpose_swap(%sparse_tensor: tensor<?x?xf64, #CSR64>) -> tensor<?x?xf64, #CSR64> {
+    %answer = graphblas.transpose %sparse_tensor { swap_sizes = true } : tensor<?x?xf64, #CSR64> to tensor<?x?xf64, #CSR64>
+    return %answer : tensor<?x?xf64, #CSR64>
+}
```
Comment (paul-tqh-nguyen) [inline mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir:87]: Not sure if this would be strictly necessary, but it might be a good idea for the sake of completeness to have a test similar to [this one](https://github.com/metagraph-dev/mlir-graphblas/blob/a6c254855fab4c0da3d08bedb2b15fcc1df98bf7/mlir_graphblas/src/test/GraphBLAS/lower_matrix_reduce_to_scalar.mlir#L25) where we test that it works with all the element types.

**Review by paul-tqh-nguyen — CHANGES_REQUESTED** [review]

File: mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h (hunk 1 above)
Comment (seibert) [reply mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h:9]: Old C++ habit.  Maybe this is not true anymore, but I thought that putting `using` statements into headers was frowned upon because it implicitly opts every .cpp file including that header (including external projects) into also automatically searching that namespace.

File: mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp (hunk 4 above)
Comment (seibert) [reply mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp:49]: Good point.  I made this change to the pass but not here.  Will fix.

File: mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir (hunk 5 above)
Comment (seibert) [reply mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir:10]: This test was generated using the utility script in the MLIR source tree and the previously tested Jinja code from functions.py.  The goal of this test was to confirm that the C++ builder code produced something structurally identical to the hand-written MLIR we had before.  (This actually caught several transcription bugs.)  I agree this test is fragile, and maybe should be deleted eventually.

File: mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir (hunk 6 above)
Comment (seibert) [reply mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir:87]: OK, I'll take a look at that test as well.

File: mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp (hunk 3 above)
Comment (seibert) [reply mlir_graphblas/src/lib/GraphBLAS/GraphBLASUtils.cpp:32]: I will make this change as well.

File: mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h (hunk 1 above)
Comment (paul-tqh-nguyen) [reply mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h:9]: I think you're right. That might be a bad idea. Maybe it'd be a better idea to do something like [this](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensor.h#L27):
```
namespace mlir {
   // insert declarations here
} // namespace mlir
```

File: mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir (hunk 6 above)
Comment (paul-tqh-nguyen) [reply mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir:87]: This might take some time and effort, so it wouldn't be a bad idea to leave a TODO comment for it.

File: mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h (hunk 1 above)
Comment (seibert) [reply mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h:9]: That also has the effect of putting these new functions into the `mlir` namespace as well, which we also probably don't want to do since these functions aren't part of MLIR.

File: mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir (hunk 6 above)
Comment (seibert) [reply mlir_graphblas/src/test/GraphBLAS/lower_transpose_swap.mlir:87]: Good idea.  Let's address this with #66.

File: mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h (hunk 1 above)
Comment (seibert) [reply mlir_graphblas/src/include/GraphBLAS/GraphBLASUtils.h:9]: Going to leave this as-is.
